docs/development/best-practices.md

```diff
@@ -0,0 +1,139 @@
+# Best practices
+
+This document explains our best practices.
+Follow these best practices when you're working on our code.
+
+## General
+
+- Prefer full function declarations for readability and better stack traces, so avoid `const func = ():void => {}`
+- Avoid [Enums](https://github.com/renovatebot/renovate/issues/13743), use union or [immutable objects](https://github.com/renovatebot/renovate/blob/5043379847818ac1fa71ff69c098451975e95710/lib/modules/versioning/pep440/range.ts#L8-L20) instead
+- Always add unit tests for full code coverage
+  - Only use `istanbul` comments for unreachable code coverage that is needed for `codecov` completion
+  - Use descriptive `istanbul` comments
+
+```js
+// istanbul ignore next: can never happen
+```
+
+### Logging
+
+Use logger metadata if logging for `WARN`, `ERROR`, `FATAL`, or if the result is a complex metadata object needing a multiple-line pretty stringification.
+Otherwise, inline metadata into the log message if logging at `INFO` or below, or if the metadata object is complex.
+
+`WARN` and above messages are often used in metrics or error catching services, and should have a consistent `msg` component so that they will be automatically grouped/associated together.
+Metadata which is separate from its message is harder for human readability, so try to combine it in the message unless it's too complex to do so.
+
+Good:
+
+```js
+logger.debug({ config }, 'Full config');
+logger.warn({ presetName }, 'Failed to look up preset');
+```
+
+Avoid:
+
+```js
+logger.debug({ branchName }, 'Generated branchName');
+logger.warn(`Failed to look up preset ${presetName}`);
+```
+
+### Array constructor
+
+Avoid the `Array()` constructor, with or without `new`, in your TypeScript code.
+It has confusing and contradictory usage.
+So you should avoid:
+
+```js
+const a = new Array(2); // [undefined, undefined]
+const b = new Array(2, 3); // [2, 3];
+```
+
+Instead, always use bracket notation to initialize arrays, or `from` to initialize an Array with a certain size i.e.
+
+```js
+// [0, 0, 0, 0, 0]
+Array.from < number > { length: 5 }.fill(0);
+```
```

Comment:
````suggestion
```ts
// [0, 0, 0, 0, 0]
Array.from<number>({ length: 5 }).fill(0);
```
````

is this still valid?

Reply:
good catch. yours is the valid one. 
no idea why i dropped the brackets there